src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java

```diff
@@ -280,18 +283,33 @@ else if (planKey.contains(BuildPlanType.SOLUTION.getName())) {
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<Result> updateResult(@RequestBody Result result) throws URISyntaxException {
         log.debug("REST request to update Result : {}", result);
-        Participation participation = result.getParticipation();
-        Course course = participation.getExercise().getCourse();
-        if (!userHasPermissions(course))
+        final var participation = result.getParticipation();
+        final var course = participation.getExercise().getCourse();
+        final var exercise = participation.getExercise();
+        if (!userHasPermissions(course) || !areManualResultsAllowed(exercise)) {
             return forbidden();
+        }
         if (result.getId() == null) {
             return createResult(result);
         }
+
         // have a look how quiz-exercise handles this case with the contained questions
         resultRepository.save(result);
         return ResponseEntity.ok().headers(HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, result.getId().toString())).body(result);
     }
 
+    private boolean areManualResultsAllowed(final Exercise exerciseToBeChecked) {
+        // Only allow manual results for programming exercises if option was enabled and due dates have passed
+        if (exerciseToBeChecked instanceof ProgrammingExercise) {
+            final var exercise = (ProgrammingExercise) exerciseToBeChecked;
+            final var relevantDueDate = exercise.getBuildAndTestStudentSubmissionsAfterDueDate() != null ? exercise.getBuildAndTestStudentSubmissionsAfterDueDate()
+                    : exercise.getDueDate();
```

Comment:
what happens if due date is also null? I would prefer that we check this, even if this should not be the case when AssessmentType is SEMI_AUTOMATIC. However we never know if this might change in the future

Reply:
I added a null check

Reply:
I think you misunderstood me. relevantDueDate could still be null (for whatever reason). Then line 307 would lead to a NullPointerException. I would say: if both `exercise.getBuildAndTestStudentSubmissionsAfterDueDate` and `exercise.getDueDate()` are null, then manual reviews are not allowed.